File: Solutions/Python/rotate_image.py
"""
    [MEDIUM]
    48. Rotate Image

    Concepts:
    - matrix
    - math

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 16.71 MB  [Beats 86.61%]
"""

class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        rot_mat = [[None]*n for _ in range(n)]

        for i in range(n):
            for j in range(n):
                rot_mat[j][i] = matrix[i][j]

        for i in range(n):
            for j in range(n):
                matrix[i][j] = rot_mat[i][n-j-1]

File: Solutions/Cpp/rotate_image.cpp
/*
    [MEDIUM]
    48. Rotate Image

    Concepts:
    - matrix
    - math

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 10.34 MB  [Beats 8.64%]
*/

class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix.size();
        std::vector<std::vector<int>> rot_mat(n, std::vector(n, -1));

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++)
                rot_mat[j][n-i-1] = matrix[i][j];
        }

        matrix = rot_mat;
    }
};